### DISCUSSION TITLE
jsPsych+JATOS saving data trial by trial

### DISCUSSION THREAD

**Discussion by CGaele** [jspsych/jsPsych#811] [Q&A] [answered]
Hi,

I'm new to jsPsych. Apologies if this question has already been answered.

Is it possible to permanently save the data (to JATOS) trial by trial rather than at the end of the experiment only? More generally, if the experiment crashes halfway through for some reason, is it possible to recover the data obtained up until that point?

Many thanks,
C

**Comment by kristian-lange** [comment]
JATOS itself allows to save data trial by trial. With jatos.appendResultData one can append data multiple time as long as the study is still running (http://www.jatos.org/jatos.js-Reference.html#jatosappendresultdata). Although I'm not so much of an expert in jsPsych, as far as I know jsPsych delivers all result data at once in the end of the whole study (but I'm happy to be corrected).

**Comment by becky-gilbert** [comment, marked as answer]
Thanks @kristian-lange! Just to add to his response: in jsPsych, it is possible to send the data back after every trial using the trial's on_finish function. It would look something like this:
```javascript
var trial = {
  ...
  on_finish: function(data) {
    // convert the data object to a JSON string
    var trial_json_results = JSON.stringify(data);
    // append this data to the end of the current results in JATOS
    jatos.appendResultData(trial_json_results);
  }
};
```
However I haven't actually tested this code so not 100% sure it works.

If you want to send the data back after _every_ trial in the experiment, you could instead put the above code to save the data into jsPsych's `on_trial_finish` parameter in the jsPsych.init options. See here: https://www.jspsych.org/core_library/jspsych-core/#jspsychinit and here: https://www.jspsych.org/overview/callbacks/#on_trial_finish

Finally, you asked about saving the data if the experiment crashes part-way through. Unfortunately it's not possible to do this if the experiment crashes due to an error on the page, but you can save the data before the participant leaves the page by putting the above code into jsPsych's `on_close` parameter, which is also part of the jsPsych.init options, and see here: https://www.jspsych.org/overview/callbacks/#on_close In this case, you'd probably want to send all of the data back to JATOS, and this will just overwrite any trial data you've already sent back.

```javascript
jsPsych.init({
  ...
  on_close: function() {
    // get all of the data and convert it to a JSON string
    var all_results = jsPsych.data.get().json();
    // send back to JATOS
    jatos.submitResultData(all_results);
  }
});
```

**Reply by jodeleeuw** [reply]
There's also the `on_data_update` event handler in `jsPsych.init({})` which can do something similar:

```js
jsPsych.init({
	...,
	on_data_update: function(data){
		var trial_json_result = JSON.stringify(data);
		jatos.appendResultData(trial_json_results);
	}
});
```

The advantage of using `on_data_update` is that it will capture every trial. The `on_finish` event would need to be added to each trial in the timeline.

**Reply by becky-gilbert** [reply]
Ah, sorry @jodeleeuw, I think I was editing my answer while you were responding because I'd forgotten about all of the callback options inside `jsPsych.init`! So my answer now mentions `on_trial_finish`. 

I think the only difference between `on_trial_finish` and `on_data_update` is that the latter function runs last (after the trial's `on_finish` and the experiment's `on_trial_finish` functions), but both should work. Is that right?

**Reply by jodeleeuw** [reply]
Yes, I think both options should work. I think the only reason to prefer `on_data_update` is that if the data were modified by `on_trial_finish` then the data could be out of sync. But that'd be a pretty rare use case!
